### PR TITLE
server: add smoke test for semi-dedicated SQL server 

### PR DESCRIPTION
**Pull request by tbg**
First commit is #47517, please don't review here. I hope it makes it through bors soon.

----

Add a test that starts up a semi-dedicated SQL server and runs some smoke test
queries. The SQL server shares some components, notably Gossip, with a test
server serving as a KV backend. The goal is to use this test as a
starting point for narrowing down and enumerating the unwanted dependencies. In
the end, the SQL server in this test should not depend on a Gossip instance
and must not rely on having a NodeID/NodeDescriptor/NodeLiveness/...

In short, it should not rely on the test server through anything other than a
`*kv.DB` and a small number of whitelisted RPCs. Nontrivial work will be
needed to get there, such as avoiding the reliance on system config
gossip.

cc @nvanbenschoten

Release note: None

### CODE REVIEW

**Review by andreimatei — APPROVED** [review]
LGTM but see the comment about the duplicate node liveness arg.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47570#-:-M53kNDm1nfwE2N00c0Z:b-oasi5r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/server/grpc_server.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47570#-M53hpCU3lu8wsMq8XW_:-M53hpCV65O4sztqkJQd:bpi3itl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3eac8b1f8df3a635eba3225efb84f8d4080c10/pkg/server/grpc_server.go#L24)):*
> ```Go
> )
> 
> // grpcServerOrNil is a wrapper on top of a grpc.Server that includes an interceptor
> ```

bad update
___
*[pkg/server/server.go, line 520 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47570#-M53hrNS41Ckdy2U9qFu:-M53hrNS41Ckdy2U9qFv:bdttu5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3eac8b1f8df3a635eba3225efb84f8d4080c10/pkg/server/server.go#L520)):*
> ```Go
> 		gossip:                   g,
> 		nodeDialer:               nodeDialer,
> 		grpcServerOrNil:          grpcServer.Server,
> ```

I for one would much rather have a comment on the field that it can be nil in tests (or whatever)
___
*[pkg/server/server_sql.go, line 106 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47570#-M53i2dgEOM3q4hPLPUW:-M53i2dgEOM3q4hPLPUX:bibz4sv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3eac8b1f8df3a635eba3225efb84f8d4080c10/pkg/server/server_sql.go#L106)):*
> ```Go
> 	status *statusServer
> 	// The DistSQLPlanner uses node liveness.
> 	nodeLiveness interface {
> ```

pls mention the production implementation in the comment
___
*[pkg/server/server_sql.go, line 119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47570#-M53i8MJ8cDqf794Ra4y:-M53i8MJ8cDqf794Ra4z:b8653du) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3eac8b1f8df3a635eba3225efb84f8d4080c10/pkg/server/server_sql.go#L119)):*
> ```Go
> 	// To register blob and DistSQL servers. This is not available if a pure SQL
> 	// server is being started, in which case we only have a Postgres listener.
> 	grpcServerOrNil *grpc.Server
> ```

same - I don't like the name :P
___
*[pkg/server/server_sql_test.go, line 96 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47570#-M53k1VO-3cjEmPutt3w:-M53k1VO-3cjEmPutt3x:b-vikyg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3eac8b1f8df3a635eba3225efb84f8d4080c10/pkg/server/server_sql_test.go#L96)):*
> ```Go
> 		args.stopper,
> 		args.Config.TestingKnobs,
> 		nl,
> ```

I'm confused. The nodeLiveness is passed both to `newSQLServer()` (through `args`), and to start? Can you at least use `args.nodeLiveness` here?

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47570)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/server/server_sql_test.go, line 45 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47570#-M53mk820lWGhYx6tsDM:-M53mk820lWGhYx6tsDN:b-obm3h6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3eac8b1f8df3a635eba3225efb84f8d4080c10/pkg/server/server_sql_test.go#L45)):*
> ```Go
> 
> // TestSQLServer starts up a semi-dedicated SQL server and runs some smoke test
> // queries. The SQL server shares some components, notably Gossip, with a test
> ```

Would it be possible to mock out the Gossip dependency at all?
There's also [`TestConn`](https://github.com/cockroachdb/cockroach/blob/27017234e8e679a035659f918f009c1942a4b25a/pkg/sql/pgwire/conn_test.go#L67) which would benefit from using just a `sqlServer` instead of the full blown dummy `TestServer` that it uses.

**Review by tbg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47570)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @andreimatei)

___
*[pkg/server/grpc_server.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47570#-M53hpCU3lu8wsMq8XW_:-M56Fby0Csrn_nMcb86F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3eac8b1f8df3a635eba3225efb84f8d4080c10/pkg/server/grpc_server.go#L24)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

bad update
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 520 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47570#-M53hrNS41Ckdy2U9qFu:-M56FcR81azL-zBrG8zq:b-u1frqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3eac8b1f8df3a635eba3225efb84f8d4080c10/pkg/server/server.go#L520)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I for one would much rather have a comment on the field that it can be nil in tests (or whatever)
</blockquote></details>

I consider that strictly a worse solution. These comments will be missed and we'll get crashes because the expectation is that an initialized struct will be fully populated.
But I realized it's really easy to make a dummy grpc server in my test, so I did that and backed this part of the change out in turn.
___
*[pkg/server/server_sql.go, line 106 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47570#-M53i2dgEOM3q4hPLPUW:-M56Gk923cLV5CpGh-eu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3eac8b1f8df3a635eba3225efb84f8d4080c10/pkg/server/server_sql.go#L106)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

pls mention the production implementation in the comment
</blockquote></details>

Done.
___
*[pkg/server/server_sql.go, line 119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47570#-M53i8MJ8cDqf794Ra4y:-M56Gkj01d4DMhky05WH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3eac8b1f8df3a635eba3225efb84f8d4080c10/pkg/server/server_sql.go#L119)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

same - I don't like the name :P
</blockquote></details>

Done.
___
*[pkg/server/server_sql_test.go, line 45 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47570#-M53mk820lWGhYx6tsDM:-M56GlPK5-Ezw6Uksa2z:bbdcahu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3eac8b1f8df3a635eba3225efb84f8d4080c10/pkg/server/server_sql_test.go#L45)):*
Anything is possible! But I don't think it's worth it. I'll make as many uses of gossip optional and then it's up to @ajwerner to do the heavy lift of ridding us of the system config gossip :muscle: 

> There's also TestConn 

It's too early to use this fabulous new sql server elsewhere. I want to clean it up some first, and even then I don't think I should be spending my time cleaning up tests that work just fine and are not in my way. But I agree with the sentiment, hopefully this sql server will become first class soon.
___
*[pkg/server/server_sql_test.go, line 96 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47570#-M53k1VO-3cjEmPutt3w:-M56I7Bi5AJeoTkbBYkT:b87palb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3eac8b1f8df3a635eba3225efb84f8d4080c10/pkg/server/server_sql_test.go#L96)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm confused. The nodeLiveness is passed both to `newSQLServer()` (through `args`), and to start? Can you at least use `args.nodeLiveness` here?
</blockquote></details>

I missed this duplication. It was a little awkward to solve, but is solved now.

**Comment by tbg** [comment]
bors r=andreimatei
